### DISCUSSION TITLE
802.11 Throughput - Showcase

### DISCUSSION THREAD

**Discussion by rhornig** [inet-framework/inet#588] [Showcases]
Discussion about the 802.11 Throughput showcase

**Comment by ghost** [comment]
When using the 802.11 Throughput showcase, attempting the switch the op mode from g to n, throws an error mentioning bitrate, however I can't seem to find the field where I can change this, any ideas?
![issue](https://user-images.githubusercontent.com/104488772/165463017-7ff5d690-722c-4f45-9d73-630040f06a7f.JPG)
?

**Comment by levy** [comment]
I guess there's no 54Mbps (which is the default datarate) in 802.11n. You need to change that parameter.

**Reply by ghost** [reply]
Even when increasing or decreasing the datarate the same error occurs, tried higher and lower values but the bitrate is always unknown, all datarates work in g, but none seem to work in n, wondering if there's a different config I need to change

**Comment by levy** [comment]
Some datarates work from here: https://en.wikipedia.org/wiki/IEEE_802.11n-2009#Data_rates
For example:
```
*.*Host.wlan[*].opMode = "n(mixed-2.4Ghz)"
*.*Host.wlan[*].bitrate = ${bitrate = 13}Mbps
```
some others are not found, some are found but lead to a runtime error.

**Reply by ghost** [reply]
Ah you are an absolute life saver! Thank you! That was causing a lot of frustration!

**Comment by levy** [comment]
These are the modes that are available (data rate, number of antennas):
```
6.5 Mbps 1
13 Mbps 1
19.5 Mbps 1
26 Mbps 1
39 Mbps 1
52 Mbps 1
58.5 Mbps 1
65 Mbps 1
14.4444 Mbps 2
28.8889 Mbps 2
43.3333 Mbps 2
57.7778 Mbps 2
86.6667 Mbps 2
115.556 Mbps 2
130 Mbps 2
144.444 Mbps 2
21.6667 Mbps 3
43.3333 Mbps 3
65 Mbps 3
86.6667 Mbps 3
130 Mbps 3
173.333 Mbps 3
195 Mbps 3
216.667 Mbps 3
28.8889 Mbps 4
57.7778 Mbps 4
86.6667 Mbps 4
115.556 Mbps 4
173.333 Mbps 4
231.111 Mbps 4
260 Mbps 4
288.889 Mbps 4
15 Mbps 1
30 Mbps 1
45 Mbps 1
60 Mbps 1
90 Mbps 1
120 Mbps 1
135 Mbps 1
150 Mbps 1
30 Mbps 2
60 Mbps 2
90 Mbps 2
120 Mbps 2
180 Mbps 2
240 Mbps 2
270 Mbps 2
300 Mbps 2
45 Mbps 3
90 Mbps 3
135 Mbps 3
180 Mbps 3
270 Mbps 3
360 Mbps 3
405 Mbps 3
450 Mbps 3
60 Mbps 4
120 Mbps 4
180 Mbps 4
240 Mbps 4
360 Mbps 4
480 Mbps 4
540 Mbps 4
600 Mbps 4
```

**Comment by levy** [comment]
Some additional info:
```
*.*Host.wlan[*].opMode = "n(mixed-2.4Ghz)"
*.*Host.wlan[*].radio.antenna.numAntennas = 4 # must be set to the number of spatial streams in the mode
*.*Host.wlan[*].radio.*.bandwidth = 40MHz # must be set to the bandwidth of the mode
*.*Host.wlan[*].bitrate = ${bitrate = 600}Mbps

#*.*Host.wlan[*].radio.separateReceptionParts = true # commented out from the example, because it causes runtime errors
#*.*Host.wlan[*].radio.separateTransmissionParts = true # commented out from the example, because it causes runtime errors
```

**Comment by solobearchn** [comment]
I was not able to reproduce the results given in the showcase, even use the same code without change.
Using 54Mbps + 2268B, the actual app layer throughput is only about 4.5Mbps, which is even not matching any cases in https://inet.omnetpp.org/docs/showcases/wireless/throughput/doc/index.html#results

<img width="545" alt="image" src="https://github.com/user-attachments/assets/59a6757c-a97d-4507-8b8c-bff7ca0d5570" />

**Reply by adamgeorge309** [reply]
What OMNeT++ and INET version are you using? Also, is the throughput correct in the .anf file? It might be a problem with the linear gauge figure.

**Reply by solobearchn** [reply]
I'm using OMNeT++ Version: 6.1, INET 4.5.4. The .anf has the correct result. 
It is not just a gauge UI display issue. I changed to use counter and the tput is 4 as shown below:
<img width="384" alt="image" src="https://github.com/user-attachments/assets/791e1dae-2fe8-475d-a6df-0c91a74d6070" />.
Seems this is a convert result between Bytes and bits which is divided by 8x:
destinationHost.app[0].packetReceived is in Bytes (8bit), so the NED code should be
`@statistic[throughput](source=liveThroughput(destinationHost.app[0].packetReceived)*8/1000000; `

**Reply by adamgeorge309** [reply]
It works for me using omnetpp 6.1 and inet-master without any modification.

![image](https://github.com/user-attachments/assets/91dcb8b8-a6fd-42ff-be13-f8a5f3bafd42)

However, in inet-4.5.4, the gauge displays the incorrect value.

**Reply by adamgeorge309** [reply]
This is fixed in the INET master and v4.5.x branches.

Here is the fix in ResultRecorders.cc: 

```
void LiveThroughputFilter::receiveSignal(cResultFilter *prev, simtime_t_cref t, cObject *object, cObject *details)
 {
     if (auto packet = dynamic_cast<cPacket *>(object))
-        receiveSignal(prev, t, packet->getByteLength(), details);
+        receiveSignal(prev, t, packet->getBitLength(), details);
 }
```

(3dc31434748b3874b8b24dc12f252053faa70c4e)
